### DISCUSSION TITLE
How to detect environment getInitialProps is running in

### DISCUSSION THREAD

**Discussion by erikdstock** [vercel/next.js#10761] [Help]
I'm trying to understand how `getInitialProps` runs in different environments: On the client and server, as well as during static export. It seems you can only infer these states from other things about the `getInitialProps` context argument. I realize that #9524 will change some of this, but want to understand this piece of the api as well as possible. Is this pretty close? 
```jsx
// pages/foo/[foo].js

Foo.getInitialProps = context => {
  const { query, req } = context;
  console.warn("GIP starting", {
    query,
    reqPresence: req && Object.keys(req).length
  });
  console.log(Object.keys(context));
  if (req) console.log(Object.keys(req));

  if (process.browser) {
    console.warn("running in browser")
    if (req && Object.keys(req).length === 0) {
      /* empty req object => statically exported */
      console.warn("statically exported")
    } else {
      /* runs after clicking <Link /> or router.push() */
      console.warn("running on Link/router.push()")
    }
  } else {
    console.warn("running on server")
    /* absence of process.browser => node process on server */

    if (Object.keys(context.req).length === 7) {
      /* Static Export
          7 static export req object has keys: 
         'url', 'headers', 'getHeader', 'setHeader',
         'hasHeader', 'removeHeader', 'getHeaderNames'
         _is this the best way to know this?_
      */

      console.warn("export");
    } else  if (Object.keys(context.req).length > 7) {
      /* req object has > 7 keys => server (SSR/SSG) */

      console.warn("SSR");
    } else {
    throw new Error("this can never happen");
  }

  return { foo: query.foo };
};

```

**Comment by erikdstock** [comment]
Hi - I realize this question might not get an answer, but wondering if I phrased it wrong or am not 'getting' something about nextjs. This static method appears key to how the entire framework works and yet the documentation around how it works and when/where it runs is sparse- possibly exacerbated by my deploying on zeit which appears to be running server-size even when i think i've done a 'static export.'

All of which is to say, I'm confused here, and perhaps I'm overcomplicating it, but if there is something I can do to improve on this question i'd appreciate it.

**Comment by willemliu** [comment]
If `req` is set then you'll know it's called server-side.

**Comment by erikdstock** [comment]
> If req is set then you'll know it's called server-side.
--
I'm aware of this but this is still just relies on inferring from other properties. Maybe my original question would be better-phrased as an RFC for a new property in the `getInitialProps` ctx argument, something like
```js
MyPage.getInitialProps = ({buildEnv, query}) => {
  switch (buildEnv) {
    case 'STATIC_EXPORT`:
      return fetchStaticExportProps(query)
    case 'SERVER':
    case 'BROWSER':
      return fetchPropsTheOtherWay(query)
  }
}
```

My sense is that folks don't need something like this because you don't build for both situations, but I find the types of checks I wrote above while simply trying to understand it extremely counterintuitive.

**Reply by willemliu** [reply]
I think you make a good point. It's convenient if the environment is exposed as a variable in the context. I just can't really think of a situation where I need to know more than checking if `req` is available or not.
